### PR TITLE
Add typings for evaluation_loop.py and remove some dead code

### PR DESCRIPTION
**Pull request by ethanwharris**
## What does this PR do?

Docs

## Before submitting
- [NA] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [NA] Did you make sure to update the documentation with your changes? (if necessary)
- [NA] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [NA] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/evaluation_loop.py [hunk 1]
```diff
@@ -11,31 +11,34 @@
...
 from pytorch_lightning.utilities.model_helpers import is_overridden
 from pytorch_lightning.utilities.signature_utils import is_param_in_hook_signature
 from pytorch_lightning.utilities.warnings import WarningCache
 
 
+if TYPE_CHECKING:
+    from torch.utils.data import DataLoader
+
+    from pytorch_lightning import Trainer
+    from pytorch_lightning.accelerators.accelerator import _STEP_OUTPUT_TYPE
```
Comment (carmocca) [inline pytorch_lightning/trainer/evaluation_loop.py:28]: Should we start putting types used throughout the codebase in a types file?

**Review by carmocca — COMMENTED** [review]
Can you run pre-commit?

`pre-commit install`
`pre-commit run --all-files` or `pre-commit run` (if changes are staged)

**Comment by ethanwharris** [comment]
@carmocca Have run pre-commit, __many__ changes haha

File: pytorch_lightning/trainer/evaluation_loop.py (hunk 1 above)
Comment (ethanwharris) [reply pytorch_lightning/trainer/evaluation_loop.py:27]: Yeah, and then only import if `TYPE_CHECKING`

Comment (carmocca) [reply pytorch_lightning/trainer/evaluation_loop.py:27]: yeah, sounds good!

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/trainer/evaluation_loop.py (hunk 1 above)
Comment (ethanwharris) [reply pytorch_lightning/trainer/evaluation_loop.py:27]: Cool, will do separate PR :smiley:

File: tests/plugins/test_ddp_plugin_with_comm_hook.py [hunk 2]
```diff
@@ -129,6 +114,4 @@ def test_ddp_spawn_fp16_compress_comm_hook(tmpdir):
         fast_dev_run=True,
     )
     trainer.fit(model)
-    assert (
-        trainer.state == TrainerState.FINISHED
-    ), f"Training failed with {trainer.state}"
+    assert (trainer.state == TrainerState.FINISHED), f"Training failed with {trainer.state}"
```
Comment (Borda) [inline tests/plugins/test_ddp_plugin_with_comm_hook.py:117]:
```suggestion
    assert trainer.state == TrainerState.FINISHED, f"Training failed with {trainer.state}"
```

File: tests/plugins/test_ddp_plugin_with_comm_hook.py [hunk 3]
```diff
@@ -100,16 +91,10 @@ def test_ddp_fp16_compress_wrap_sgd_comm_hook(tmpdir):
...
-    expected_comm_hook = default.fp16_compress_wrapper(
-        powerSGD.powerSGD_hook
-    ).__qualname__
+    trainer_comm_hook = (trainer.accelerator.training_type_plugin._model.get_ddp_logging_data().comm_hook)
+    expected_comm_hook = default.fp16_compress_wrapper(powerSGD.powerSGD_hook).__qualname__
     assert trainer_comm_hook == expected_comm_hook
-    assert (
-        trainer.state == TrainerState.FINISHED
-    ), f"Training failed with {trainer.state}"
+    assert (trainer.state == TrainerState.FINISHED), f"Training failed with {trainer.state}"
```
Comment (Borda) [inline tests/plugins/test_ddp_plugin_with_comm_hook.py:97]:
```suggestion
    assert trainer.state == TrainerState.FINISHED, f"Training failed with {trainer.state}"
```

File: tests/plugins/test_ddp_plugin_with_comm_hook.py [hunk 4]
```diff
@@ -71,14 +66,10 @@ def test_ddp_sgd_comm_hook(tmpdir):
...
-    trainer_comm_hook = (
-        trainer.accelerator.training_type_plugin._model.get_ddp_logging_data().comm_hook
-    )
+    trainer_comm_hook = (trainer.accelerator.training_type_plugin._model.get_ddp_logging_data().comm_hook)
     expected_comm_hook = powerSGD.powerSGD_hook.__qualname__
     assert trainer_comm_hook == expected_comm_hook
-    assert (
-        trainer.state == TrainerState.FINISHED
-    ), f"Training failed with {trainer.state}"
+    assert (trainer.state == TrainerState.FINISHED), f"Training failed with {trainer.state}"
```
Comment (Borda) [inline tests/plugins/test_ddp_plugin_with_comm_hook.py:72]:
```suggestion
    assert trainer.state == TrainerState.FINISHED, f"Training failed with {trainer.state}"
```

File: tests/plugins/test_ddp_plugin_with_comm_hook.py [hunk 5]
```diff
@@ -43,14 +42,10 @@ def test_ddp_fp16_compress_comm_hook(tmpdir):
...
-    trainer_comm_hook = (
-        trainer.accelerator.training_type_plugin._model.get_ddp_logging_data().comm_hook
-    )
+    trainer_comm_hook = (trainer.accelerator.training_type_plugin._model.get_ddp_logging_data().comm_hook)
     expected_comm_hook = default.fp16_compress_hook.__qualname__
     assert trainer_comm_hook == expected_comm_hook
-    assert (
-        trainer.state == TrainerState.FINISHED
-    ), f"Training failed with {trainer.state}"
+    assert (trainer.state == TrainerState.FINISHED), f"Training failed with {trainer.state}"
```
Comment (Borda) [inline tests/plugins/test_ddp_plugin_with_comm_hook.py:48]:
```suggestion
    assert trainer.state == TrainerState.FINISHED, f"Training failed with {trainer.state}"
```

**Review by Borda — COMMENTED** [review]
if it is fine for the file, remove it from ignore list...

**Review by Borda — APPROVED** [review]

**Review by SeanNaren — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
Awesome work !
